sky/skylet/providers/ibm/node_provider.py

```diff
@@ -0,0 +1,924 @@
+        self.endpoint = f"https://{self.region}.iaas.cloud.ibm.com"
+        self.zone = self.provider_config["availability_zone"]
+        self.ibm_vpc_client = _get_vpc_client(
+            self.endpoint, IAMAuthenticator(
+                self.iam_api_key, url=self.iam_endpoint)
+        )
+        self.vpc_tags = None
+        self.vpc_provider = IBMVPCProvider(
+            self.resource_group_id, self.region, self.cluster_name)
+
+        self._load_tags()
+
+        # Cache of starting/running/pending(below PENDING_TIMEOUT)nodes.
+        #                               in the format{node_id:node_data}.
+        self.cached_nodes = {}
+        # cache of the nodes created, but not yet tagged and running.
+        #                    in the format {node_id:time_of_creation}.
+        self.pending_nodes = {}
+        self.deleted_nodes = []  # ids of nodes scheduled for deletion.
+
+        # if cache_stopped_nodes == true, nodes will be stopped
+        #   instead of deleted to better accommodate future rise in demand
+        self.cache_stopped_nodes = provider_config.get(
+            "cache_stopped_nodes", True)
+
+    def _get_node_type(self, name):
+        if f"{self.cluster_name}-{NODE_KIND_WORKER}" in name:
+            return NODE_KIND_WORKER
+        elif f"{self.cluster_name}-{NODE_KIND_HEAD}" in name:
+            return NODE_KIND_HEAD
+
+    def _get_nodes_by_tags(self, filters):
+        """
+        returns list of nodes who's tags are matching the specified filters. 
+        Args:
+            filters(dict):  specified conditions to filter nodes by.
+        """
+
+        nodes = []
+        # no filters were specified or the only filter is the type of the node
+        if (not filters or
+                list(filters.keys()) == [TAG_RAY_NODE_KIND]):
+            result = self.ibm_vpc_client.list_instances().get_result()
+            instances = result["instances"]
+            while result.get("next"):
+                start = result["next"]["href"].split("start=")[1]
+                result = self.ibm_vpc_client.list_instances(
+                    start=start).get_result()
+                instances.extend(result["instances"])
+
+            for instance in instances:
+                kind = self._get_node_type(instance["name"])
+                if kind and instance["id"] not in self.deleted_nodes:
+                    if not filters or kind == filters[TAG_RAY_NODE_KIND]:
+                        nodes.append(instance)
+                        with self.lock:
+                            node_cache = self.nodes_tags.setdefault(
+                                instance["id"], {})
+                            node_cache.update(
+                                {
+                                    TAG_RAY_CLUSTER_NAME: self.cluster_name,
+                                    TAG_RAY_NODE_KIND: kind,
+                                }
+                            )
+        else:  # match filters specified
```

Comment:
Haven't forget this one. I'm not sure why we (me&my predecessor) didn't do as in aws's implementation and list all the instances, update, and then split into cases...
Trying to figure out if there's some deep understanding of Ray behind this decision. 
Will get back to you on this soon.

Reply:
o.k, took a deep dive into this one and this I believe i understand the logic behind it: 
- our `self.node_tags` is loaded from file whenever a new instance of the module is created, and the node tags (both in memory and local cache) are updated whenever a node is either created/deleted. 
- if some error occurred in between the time the in-memory cache and the local cache was updated (as might also happen on the other providers for example: error in between the time the in-memory cache and the tagging of the instances takes place) and the tags are basic, we can filter without the in memory cache, because we don't need its tags (we can deduce the kind of node and the cluster's name). 
However, if the specified filters are more complex, we can only have them in our memory cache - `self.nodes_tags`, i.e. we can't deduce them on our own. hence, if a node_id doesn't appear in this cache we can't tell if it fits a complicated set of filters, thus no point in updating the cache, because it already contains the correct fields and key (node_id) . 

added documentation and fixes (updates to self.node_tags didn't propagate to local cache).
Would like to get your feedback on the matter.

Reply:
Ah, I see. So the local cache file only contains basic tags?